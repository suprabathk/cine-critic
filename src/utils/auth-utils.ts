import axios from "axios";

const getAuthTokenOptions = {
  method: "GET",
  url: `${process.env.NEXT_PUBLIC_TMDB_API}/authentication/token/new`,
  headers: {
    accept: "application/json",
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_TMDB_API_READ_ACCESS_TOKEN}`,
  },
};

export const getAuthToken = async () => {
  return (await axios.request(getAuthTokenOptions)).data.request_token;
};

const getSignInOptions = (data: {
  username: string;
  password: string;
  request_token: string;
}) => ({
  method: "POST",
  url: `${process.env.NEXT_PUBLIC_TMDB_API}/authentication/token/validate_with_login`,
  headers: {
    accept: "application/json",
    "content-type": "application/json",
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_TMDB_API_READ_ACCESS_TOKEN}`,
  },
  data: data,
});

export const signIn = async (data: {
  username: string;
  password: string;
  request_token: string;
}) => {
  try {
    return await axios.request(getSignInOptions(data));
  } catch (error) {
    return { data: { success: false } };
  }
};

export const isAuth = () =>
  typeof window !== undefined && !!localStorage.getItem("authToken");
