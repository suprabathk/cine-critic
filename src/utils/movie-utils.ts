import { MovieDetails } from "@/types/movies";
import axios from "axios";

const getNowPlayingMoviesListOptions = {
  method: "GET",
  url: "https://api.themoviedb.org/3/movie/now_playing?language=en-US",
  headers: {
    accept: "application/json",
    Authorization:
      "Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiIwMDNlMmY1NTdlMjMxZmZkNmQ5ZDNjODMwM2NjYTRiMSIsInN1YiI6IjY1MjRmOGRmMGNiMzM1MTZmZDQ2ZWJiNyIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.GceWSXdFFaaI77fSU7zVhan_nRshe3c5GSGoZcVj35Y",
  },
};

export const getNowPlayingMoviesList = async () => {
  return (await axios.request(getNowPlayingMoviesListOptions)).data;
};

const getDiscoverOptions = {
  method: "GET",
  url: "https://api.themoviedb.org/3/discover/movie?sort_by=popularity.desc",
  headers: {
    accept: "application/json",
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_TMDB_API_READ_ACCESS_TOKEN}`,
  },
};

export const getDiscoverList = async () => {
  console.log("hiihi");

  return (await axios.request(getDiscoverOptions)).data;
};

const getMovieDetailsOptions = (movieID: string) => ({
  method: "GET",
  url: `https://api.themoviedb.org/3/movie/${movieID}?language=en-US`,
  headers: {
    accept: "application/json",
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_TMDB_API_READ_ACCESS_TOKEN}`,
  },
});

export const getMovieDetails = async (
  movieID: string
): Promise<MovieDetails> => {
  return (await axios.request(getMovieDetailsOptions(movieID))).data;
};
