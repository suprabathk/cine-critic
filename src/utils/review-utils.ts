import { Review } from "@/types/review";

export const getReviews = async (movieID: string): Promise<Review[]> => {
  return await Promise.resolve(
    typeof window !== "undefined" &&
      JSON.parse(localStorage.getItem(movieID) ?? "[]")
  );
};

export const addReview = async (movieID: string, review: Review) => {
  return await Promise.resolve(
    typeof window !== "undefined" &&
      localStorage.setItem(
        movieID,
        JSON.stringify([
          { ...review, rating: review.rating },
          ...JSON.parse(localStorage.getItem(movieID) ?? "[]"),
        ])
      )
  );
};

export const deleteReview = async (movieID: string, reviewID: string) => {
  const reviews: Review[] = JSON.parse(localStorage.getItem(movieID) ?? "[]");

  return await Promise.resolve(
    typeof window !== "undefined" &&
      localStorage.setItem(
        movieID,
        JSON.stringify(reviews.filter((review) => review.id !== reviewID))
      )
  );
};

export const editReview = async (movieID: string, newReview: Review) => {
  const reviews: Review[] = JSON.parse(localStorage.getItem(movieID) ?? "[]");
  return await Promise.resolve(
    typeof window !== "undefined" &&
      localStorage.setItem(
        movieID,
        JSON.stringify(
          reviews.map((review) =>
            newReview.id === review.id ? newReview : review
          )
        )
      )
  );
};
