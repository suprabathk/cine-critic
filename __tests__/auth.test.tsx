import React from "react";
import renderer from "react-test-renderer";
import SignIn from "@/pages/auth/index";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { signIn } from "@/utils/auth-utils";

jest.mock("../src/utils/auth-utils", () => ({
  getAuthToken: jest.fn().mockResolvedValue(""),
  isAuth: jest.fn().mockReturnValue(true),
  signIn: jest.fn().mockResolvedValue({ data: { success: true } }),
}));

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: { movieID: "" },
      push: jest.fn(),
    };
  },
}));

describe("Auth page tests", () => {
  it("Snapshot testing", () => {
    const tree = renderer.create(<SignIn />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Renders correctly", async () => {
    render(<SignIn />);
    const submit = screen.getByRole("button", {
      name: /sign in/i,
    });
    const usernameInput = screen.getByLabelText(/username/i);
    const passwordInput = screen.getByLabelText(/password/i);
    expect(submit).toBeInTheDocument();
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
  });

  describe("Username tests", () => {
    it("Username is required", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(screen.queryByText(/username is required/i)).toBeNull();

      await userEvent.click(submit);
      //Calling fucntion after click
      await waitFor(() => {
        expect(screen.queryByText(/username is required/i)).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });

    it("Username should be atleast 8 characters long", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(
        screen.queryByText(/username should be minimum of 8 characters/i)
      ).toBeNull();
      const usernameInput = screen.getByLabelText(/username/i);
      await userEvent.type(usernameInput, "123456");

      await userEvent.click(submit);

      //Errors after click
      await waitFor(() => {
        expect(
          screen.queryByText(/username should be minimum of 8 characters/i)
        ).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });

    it("Username cannot contain special symbols", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(
        screen.queryByText(/username can contain only letters and numbers/i)
      ).toBeNull();

      const usernameInput = screen.getByLabelText(/username/i);
      await userEvent.type(usernameInput, "1234567890#$");
      await userEvent.click(submit);
      //Calling fucntion after click
      await waitFor(() => {
        expect(
          screen.getByText(/username can contain only letters and numbers/i)
        ).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });
  });

  describe("Password tests", () => {
    it("Password is required", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(screen.queryByText(/password is required/i)).toBeNull();

      await userEvent.click(submit);
      //Calling fucntion after click
      await waitFor(() => {
        expect(screen.queryByText(/password is required/i)).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });

    it("Password should be atleast 8 characters long", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(
        screen.queryByText(/password should be minimum of 8 characters/i)
      ).toBeNull();
      const passwordInput = screen.getByLabelText(/password/i);
      await userEvent.type(passwordInput, "123456");

      await userEvent.click(submit);

      //Errors after click
      await waitFor(() => {
        expect(
          screen.queryByText(/password should be minimum of 8 characters/i)
        ).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });

    it("Password cannot contain spaces", async () => {
      render(<SignIn />);
      const submit = screen.getByRole("button", {
        name: /sign in/i,
      });
      //No errors initially
      expect(screen.queryByText(/password cannot contain spaces/i)).toBeNull();

      const passwordInput = screen.getByLabelText(/password/i);
      await userEvent.type(passwordInput, "123456 78");

      await userEvent.click(submit);
      //Calling fucntion after click
      await waitFor(() => {
        expect(
          screen.queryByText(/password cannot contain spaces/i)
        ).toBeInTheDocument();
        expect(signIn).toBeCalledTimes(0);
      });
    });
  });

  it("Submitting the form", async () => {
    render(<SignIn />);
    const submit = screen.getByRole("button", {
      name: /sign in/i,
    });
    //No errors initially
    const usernameInput = screen.getByLabelText(/username/i);
    const passwordInput = screen.getByLabelText(/password/i);

    await userEvent.type(usernameInput, "1234567890");
    await userEvent.type(passwordInput, "1234567890");
    await userEvent.click(submit);
    //Calling fucntion after click
    await waitFor(() => {
      expect(signIn).toBeCalledTimes(1);
    });
  });
});
