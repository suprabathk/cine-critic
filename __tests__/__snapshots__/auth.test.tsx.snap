// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth page tests Snapshot testing 1`] = `
<div
  className="MuiStack-root css-9gyggb-MuiStack-root"
>
  <div
    className="authInfo"
  >
    <div
      className="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorError MuiChip-filledError css-vmjpzh-MuiChip-root"
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <span
        className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        Movies, Shows and more
      </span>
    </div>
    <h3
      className="MuiTypography-root MuiTypography-h3 css-3sj0e9-MuiTypography-root"
    >
      Delivering movie reviews around the globe.
    </h3>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Get the live reviews and latest news about Movies, Shows.
    </p>
  </div>
  <form
    className="authForm"
    onSubmit={[Function]}
  >
    <h2
      className="authHeading"
    >
      Welcome back to 
      <br />
      <a
        className="logoText"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        CineCritic
      </a>
    </h2>
    <p
      className="authDescription"
    >
      Sign in to your account.
    </p>
    <div
      className="formInput"
    >
      <label
        className="formLabel"
        htmlFor="username"
      >
        Username
      </label>
      <input
        className="formTextInput"
        id="username"
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="JohnDoe"
        type="text"
        value=""
      />
      <span
        className="formErrorMessage"
      />
    </div>
    <div
      className="formInput"
    >
      <label
        className="formLabel"
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="formTextInput"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="•••••••••"
        type="password"
        value=""
      />
      <span
        className="formErrorMessage"
      />
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-dx9gco-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
